sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -150,28 +150,46 @@ class GpuSingleDirectoryDataWriter(
   }
 
   override def write(batch: ColumnarBatch): Unit = {
-    var needToCloseBatch = true
-    try {
-      // TODO: This does not handle batch splitting to make sure maxRecordsPerFile is not exceeded.
-      if (description.maxRecordsPerFile > 0 && recordsInFile >= description.maxRecordsPerFile) {
-        fileCounter += 1
-        assert(fileCounter < MAX_FILE_COUNTER,
-          s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
-
-        newOutputWriter()
-      }
+    withResource(batch) { batch =>
+      withResource(GpuColumnVector.from(batch)) {table => 
+        val numRows = table.getRowCount()
+        val maxRecordsPerFile = description.maxRecordsPerFile.toInt
+        val splitIndexes = (maxRecordsPerFile > 0, recordsInFile < maxRecordsPerFile) match {
+          case (false, _) => IndexedSeq.empty
+          case (true, false) =>
+            (1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => i * maxRecordsPerFile)
+          case (true, true) => {
+            val filledUp = maxRecordsPerFile - recordsInFile.toInt
+            val remain = numRows - filledUp
+            (0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
```

Comment:
Scala doesn't have the ceiling division function and we don't have this function in our utils.

I'm not sure where to put this function, so just unfold the expression here.

we could file a follow-up to clean this

Reply:
For the following case, will get split indices: (10, 20), but we actually want to get (5, 15, 25) to make sure eache file contains 10 rows.

```
var recordsInFile = 5   // The current file already have 5 rows.
var maxRecordsPerFile = 10
var numRows = 30
(1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt).map(i => maxRecordsPerFile * i)
```

Reply:
When `recordsInFile < maxRecordPerFile`, we calculate the split index by
```scala
val filledUp = maxRecordsPerFile - recordsInFile.toInt
val remain = numRows - filledUp
(0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
    .map(i => filledUp + i * maxRecordsPerFile)
```

For your example, the process is
```
val filledUp = 10 - 5 = 5;
val remain = 30 - 5 = 25;
(0 until (25 + 10 -1) / 10).map(i => 5 + i * 10)
   -->
(0, 1, 2).map(i => 5 + 10*i)
    -->
(5, 15, 25)
```

Reply:
Yeah, but do we need the `recordsInFile < maxRecordsPerFile`?  This should be always true.
`recordsInFile` indicates the row number of the current file.

Reply:
`recordesInFile` can be equal to `maxRecordsPerFile` if the last writer just fill up the file, in which case we should open a new file directly.

Reply:
Can we combine 2 branches into 1?

```
          case (true, false) =>
            (1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
              .map(i => i * maxRecordsPerFile)

          case (true, true) => {
            val filledUp = maxRecordsPerFile - recordsInFile.toInt
            val remain = numRows - filledUp
            (0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)

```

After splitting, we can call n-1 (n is the split number) times of `newOutputWriter()`, just keep the last split without `newOutputWriter()`.